WordPress/gutenberg
# Toolbar: Use Ariakit instead of Reakit

## What?


This PR updates the Toolbar component from the `@wordpress/components` package to use Ariakit instead of Reakit. [Ariakit](https://ariakit.org) is the successor of [Reakit](https://reakit.io).

## Why?


Ariakit is already superior to Reakit in all aspects. It is faster, offers more features, improvements, and has fewer bugs. In fact, dozens of bugs have been reported in the past year, and all of them could be resolved in userland before a fix was implemented in the library. See [closed bugs](https://github.com/ariakit/ariakit/issues?q=is%3Aissue+label%3A%22bug%22+is%3Aclosed+) and notice that all of them have the "has workaround" label.

Ariakit provides full support for the latest versions of React (17 and 18). It is still released as v0.x.x because we are still working on the website and preparing for an official release.

Breaking changes are released as v0.x.0, following semver. Additionally, most of these changes are preceded by soft deprecations (without warnings), followed by deprecation warnings before the actual breaking change.

## How?


I updated the Toolbar component to use the new API (mostly `state` → `store`). I've also updated the `ToolbarItem` prop types.

## Testing Instructions





The Toolbar component is mainly used in the `Block tools` toolbar (the block one) and the `Document tools` toolbar (the top one) on the editor. There should be no change in how they work.

Those components are covered by the following e2e tests:

- `navigable-toolbar.spec.js`
- `shortcut-focus-toolbar.spec.js`
- `toolbar-roving-tabindex.spec.js`

All tests should pass.

### Testing Instructions for Keyboard


The `shortcut-focus-toolbar.spec.js` and `toolbar-roving-tabindex.spec.js` tests cover most of the keyboard interactions on the toolbar.

Make sure you can navigate to the block toolbar using either <kbd>Shift</kbd><kbd>Tab</kbd> on the block or <kbd>Alt</kbd><kbd>F10</kbd>. Once in the toolbar, you should be able to navigate through the items using arrow keys and interact with buttons by pressing <kbd>Enter</kbd> and <kbd>Space</kbd>. Dropdowns should open on the <kbd>ArrowDown</kbd> key and close with <kbd>Esc</kbd>, returning the focus to the dropdown button in the toolbar.

Interactions should also remain unchanged when the `Top toolbar` setting is enabled (in the `Options` menu located at the end of the `Editor top bar` region). **Note**: As a sighted keyboard user, I noticed the behavior here is quite confusing. The visual position of the elements in the top toolbar (that now has both the block tools and document tools) doesn't seem to match the tab order. But this behavior is also present on trunk, so I guess it's out of the scope of this PR.
